drivers/include/periph/spi.h

```diff
@@ -334,41 +339,52 @@ int spi_init_with_gpio_mode(spi_t bus, spi_gpio_mode_t mode);
  * Starting a new SPI transaction will get exclusive access to the SPI bus
  * and configure it according to the given values. If another SPI transaction
  * is active when this function is called, this function will block until the
- * other transaction is complete (spi_relase was called).
- *
- * @note    This function expects the @p bus and the @p cs parameters to be
- *          valid (they are checked in spi_init and spi_init_cs before)
+ * other transaction is complete (@ref spi_release was called).
  *
- * @param[in] bus       SPI device to access
- * @param[in] cs        chip select pin/line to use, set to SPI_CS_UNDEF if chip
+ * @param[in]   bus     SPI device to access
+ * @param[in]   cs      chip select pin/line to use, set to SPI_CS_UNDEF if chip
  *                      select should not be handled by the SPI driver
- * @param[in] mode      mode to use for the new transaction
- * @param[in] clk       bus clock speed to use for the transaction
- *
- * @return              SPI_OK on success
- * @return              SPI_NOMODE if given mode is not supported
- * @return              SPI_NOCLK if given clock speed is not supported
+ * @param[in]   mode    mode to use for the new transaction
+ * @param[in]   clk     maximum bus clock to use in Hz
+ *
+ * @return  The actually used clock frequency in Hz
```

Comment:
In some case this lead into additional computation and we don't need this information every time spi_acquire is called so what do you think of a dedicated function as in #16727. Moreover this allow to retrieve the resulting speeds without providing mode and cs parameters as in the cmd_init function's help in tests/periph_spi.

Reply:
Note that the API was not only changed to return the frequency used, but also changes the way to specify frequencies (for performance reasons). As a result, the implementation will have to figure out which prescaler to use to generate the requested maximum frequency. I expect that at this point, the actually used frequency value is already available anyway without additional computation.

I believe that a dedicated function will cause more ROM being used than just changing `spi_acquire()` to just return the frequency value it has computed (for picking the right prescaler) anyway.

Reply:
On 8-bit, just comparing or returning an uint32_t leads to additional computations. That's why the atmega implementation was overriding the spi_clk_t enum with some encoded registry bits. Anyway we don't need this data every time the bus is acquired.

Reply:
Well, it is a typical trade-off. Choosing the frequency not via enum but by specifying the frequency in Hertz is the main contributor to the overhead. Returning the frequency that was selected will be almost free (even on AVR, since the compiler can just leave the selected frequency in the registers the return value is expected).

So I assume this is mostly about should we switch away from the enum to an `uint32_t`. This is however needed to support arbitrary frequencies. If we would keep using an `enum`, we will be limited to a couple of frequency options that best match the frequencies supported by most MCUs. If we would want to extend the enum, we would have to touch *every* implementation - every time we a new constant is added to the enum. Or worse: The supported constants will be so fragmented, that we get an `#if` hell in *every* driver when calling `spi_acquire()`.

I agree that the current option has less overhead, especially on AVR. But it performs a lot worse for the faster 32 bit MCUs that do offer high clock rates, since SPI transfers are slowed down *a lot*. And when looking at the activity of RIOT's AVR support that is still lacking basic features like a high-level support of PROGMEM and power management, there is not much interest in running RIOT on AVR.

Also: The current `spi_acquire()` and `spi_release()` isn't for free even as of now. When I rewrote the `cc110x` driver I was able to cut down the round trip time to one quarter of the original RTT, mainly by not acquiring and releasing the SPI bus for every single byte transferred. Powering up and initializing the SPI bus often is more costly than spending a few CPU cycles on picking the right clock, so the overhead relative to the current cost of `spi_acquire()` is not too big. And if this allows you to clock the SPI much higher, it will end up speeding up the transfer quite a bit (which in turn allows the MCU to go to lower power mode sooner).

I believe that this trade-off is immensely beneficial for all MCUs but AVR. And I believe that it is still beneficial for AVR (but admittedly the overhead of 32 bit arithmetic is quite a bit bigger there), since even there higher SPI clocks will become available with the API change. See the example given in the "increasing performance" bullet point in the PR description for why this can happen (and even is likely to happen) on slow MCUs.

Reply:
> I believe that a dedicated function will cause more ROM being used than just changing `spi_acquire()` to just return the frequency value it has computed (for picking the right prescaler) anyway.

In most cases the prescaler is directly computed so if you want to get the resulting speed, you will have to compute it from the prescaler.

In my opinion the spi_acquire() function handle too many things and have too many input parameters.

For example the cs parameter is ignored in most implementations but it can lead to think that you must release and acquire the bus every time you change cs when it is the functions spi_transfer_* that actually handles this.

There are mostly 2 kinds of prescaler hardware implementations apart from the weird one from kinetis:

1. Those that use a logic comparator at the output of a clock counter so the prescaler computation is no more than a rounded up division (cc2538, efm32, esp_common, fe310, lm4f120, lpc23xx, msp430fxyz, qn908x, sam0_common, sam3).
2. Those who use a simple bit selector so the value you need in register is log₂(prescaler), not so easy to rounding up (atmega_common, atxmega, nrf51, nrf52, stm32). See the _get_clkdiv() function in the stm32 implementation (the only one of them that supports arbitrary clock rate for now).

The stm32 implementation uses some caching in order to mitigate the overhead but this is a waste in the case you use peripherals with different clock rate on the same bus.

I think we could do this:
```c
/**
 * @brief   Compute the needed value for picking the right prescaller
 *
 * @param[in] freq      maximum bus clock to use in Hz
 *
 * @return              prescaler value or selector value or something else
 *                      depending on the hardware implementation
 */
spi_clk_t spi_clk_sel(uint32_t freq);

void spi_acquire(spi_t bus, spi_mode_t mode, spi_clk_t clk_sel);

/**
 * @brief   Get the resulting frequency
 *
 * @param[in] clk_sel   clock selector
 *
 * @return              the resulting frequency
 */
uint32_t spi_clk_freq(spi_clk_t clk_sel);
```
This way it is clear that you do not need to to release and acquire the bus every time you change cs and also that you will not necessarily obtain the frequency you expect. In addition, you can cache as many prescalers as you need in your application.

Reply:
> For example the cs parameter is ignored in most implementations

It is there for the implementations that need it. E.g. some MCUs have a dedicated hardware CS pin that is handled differently from GPIOs and require different setup. For this reason it needed to know whether the CS pin requires special handling or not.

Some MCUs allow disconnecting a GPIO pin while keeping a pull resistor connected - or even just add an external pull resistor to not have the CS line floating. In some cases, this could result in lower power consumption when the CS line is no longer actively driven while the SPI is no longer used.

> but it can lead to think that you must release and acquire the bus every time you change cs when it is the functions spi_transfer_* that actually handles this.

Well, this assumption is correct. If you switch from a regular GPIO used as CS to a magic hardware CS pin, things can go south on some implementations.

I do get your point that it would be easier to just force implementations to treat the CS pin as regular GPIO (even if it could be used as a magic hardware CS pin) and drop it from `spi_acquire()`. Likely this would simplify things and reduce ROM. But this is not really related to the clock handling of the SPI bus that this PR focuses on improving, so IMO that should be a separate PR.

> I think we could do this:

I see the point. One downside that I see is that this will not work well with dynamic CPU frequency scaling, as one would have to recompute the prescaler after every frequency change. If it is computed with every `spi_acquire()`, changing the CPU frequency would work fine whenever the SPI bus is idle.

I just asked the usual suspects in Matrix. If there is no indication that this will make it to RIOT any time soon, I will update as suggested. But I will keep the CS parameter, as this is IMO an independent discussion and unrelated to the frequency handling.

Reply:
It seems that CPU frequency scaling is not imminent and likely will only become reality for MCUs that has a peripheral clock domain that is independent of the CPU clock domain, so that changing the CPU frequency will not force recalculation of all peripheral clock calculation. Hence, I took the suggestion.

Reply:
Thanks, but why persist in returning the frequency here now that we have a dedicated function to do so?

Reply:
fixed